text/1045-recommend-arrows-instead-of-action-decorator.md

```diff
@@ -0,0 +1,117 @@
+
+
+`@action` is only _needed_ when we know that the `this` context would be lost. Historically this has been taught as "when you invoke a function from a template, or pass a function somewhere else".
+We can use that same heuristic for "when to define an arrow function", and **this should be mentioned in the guides**, along with a sample error message "can't access x on undefined" (though, I think some browsers would also print `this.x` in their error message, giving more context / hints to the fact that `this` is undefined. More sinisterily, though, the `this` could be the _wrong_ `this`. This is a hazard of working with `this` in JavaScript, and why some people liberally used `@action` on functions, so they didn't have to worry about this problem at all. Because of liberal usage of `@action`, the above-described runtime cost grows more than it needs to. Do note that this cost is small.
+
+With arrow functions, we completely eliminate all of the memory upkeep that the `@action` decorator is doing. We still duplicate the methods per-instance (those that are defined as arrows), but all the upkeep around those methods is _handled for us_ by the JavaScript engine. In modern code, using arrow functions is kind of the equivelent of only having [this line](https://github.com/emberjs/ember.js/blob/v5.11.0/packages/%40ember/object/index.ts#L215C13-L215C14) from the `@action` decorator.
+
+## How we teach this
+
+1. Update all usages of `@action` to be arrows in the guides, removing all imports of the `@action` decorator.
+2. On the ["Component State and Actions"](https://guides.emberjs.com/release/components/component-state-and-actions/) page (where action is first used), explain the purpose of the arrow function, and why we use it. (this explanation is currently missing for `@action` as well -- however at the bottom of the page, it links to ["Patterns for Actions"](https://guides.emberjs.com/release/in-depth-topics/patterns-for-actions/) -- which _also_ does not mention anything about this-binding. We can link out to MDN ([maybe this one](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/this#bound_methods_in_classes) -- or we could write something ourselves for a stable-reference to that content).  We should also explain the advantages of _not_ using an arrow function (shared memory between instances).
+
```

Comment:
A part 3 may be needed on inheritance -- which we should discourage, as misuse of inheritance is often folks mess up OOP. With components, composition is _very nice_, and with classes, dependency injection is _very nice_.

Here is what happens with arrow functions and you try to use "the same patterns" as you would with methods:
```js
class A {
  message = "hi";
  greet = () => this.message;
}


class B extends A {
  greet = () => `:::: ${super.greet()}`;
}

let b = new B();  

console.log({
  b: b.greet(),
});
```

gives the error:
```
runner-4.1.8.min.js:1 error TypeError: (intermediate value).greet is not a function
    at B.greet (<anonymous>:9:35)
    at <anonymous>:15:12
```

Reply:
This happens because instance fields are not present on the prototype at all.
You can't do this either:
```js
class B extends A {
  greet = () => `:::: ${A.prototype.greet.call(this)}`;
}
```
prototype.greet is undefined.

Reply:
If folks _want_ to use inheritance, despite warnings, this RFC doesn't deprecate `@action`

_at some point_ what `@action` is doing will _do less_ after classic paradigms are deprecated and removed

Reply:
I was about to post about this, that it breaks inheritance.

But if this RFC is --just-- about recommending something while leaving the door open to use `@action` and real class methods, then fine. I don't think we should go against this pattern (class prototype methods) even if it means keeping `@action` around for who knows how long.

Reply:
yeah, until JS figures out better ergonomics around class-method-this-binding, we should keep `@action`